### ISSUE TITLE
intermittent test failure in testCallLater

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#10726]
|[<img alt="warner's avatar" src="https://avatars.githubusercontent.com/u/27146?s=50" width="50" height="50">](https://github.com/warner)| @warner reported|
|-|-|
|Trac ID|trac#1113|
|Type|defect|
|Created|2005-07-16 08:26:24Z|

<details><summary>Searchable metadata</summary>

```
trac-id__1113 1113
type__defect defect
reporter__warner warner
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1121502384000000 1121502384000000
changetime__1156113420000000 1156113420000000
version__ 
owner__ 
cc__exarkun cc__warner
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="warner's avatar" src="https://avatars.githubusercontent.com/u/27146?s=50" width="50" height="50">](https://github.com/warner)|@warner commented|
|-|-|

```
#!html
<pre>
The buildbot is seeing a timing-related test failure in testCallLater:



[FAIL]: twisted.test.test_internet.InterfaceTestCase.testCallLater

  File
"/home/buildbot/BuildBot/slave/full2.3/Twisted/twisted/internet/defer.py", line
109, in maybeDeferred
    result = f(*args, **kw)
  File
"/home/buildbot/BuildBot/slave/full2.3/Twisted/twisted/test/test_internet.py",
line 222, in testCallLater
    self.assertApproximates(self._calledTime, start + 0.5, 0.2 )
  File
"/home/buildbot/BuildBot/slave/full2.3/Twisted/twisted/trial/unittest.py", line
418, in &lt;lambda>
    assertApproximates = lambda self, a, b, c, msg=None: assertApproximates(a,
b, c, msg)
  File
"/home/buildbot/BuildBot/slave/full2.3/Twisted/twisted/trial/assertions.py",
line 176, in assertApproximates
    raise FailTest, (msg or "%s ~== %s" % (first, second))
twisted.trial.assertions.FailTest: 1121392438.02 ~== 1121392437.74

Clearly this is a result of a heavily loaded test host being unable to complete
an operation within 200ms of the target time. Perhaps we can change the test to
assert some other property of the reactor.callLater function being exercised..
the fact that it completes at all? The fact that it does not complete *before* a
similar callLater that uses a shorter timeout? Likewise it should complete
before a callLater with a longer timeout?

There should probably still be a failsafe timeout, but it can use a much longer
value (30 seconds), because it should only ever be used when the reactor is so
horribly broken that timers fail altogether.

cheers,
 -Brian
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun removed owner|
|-|-|

Review in call-later-timing-failures-1113

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_3"></a>|@radix set owner to @exarkun|
|-|-|

I can't think of anything bad to say about this change. Maybe the docstring of the TestCase, but I agree with it, so not really bad.


Merge it quickly, before I fully understand the terror of the test that it is replacing.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set status to `closed`|
|-|-|

(In [17886]) Merge call-later-timing-failures-1113

Author: exarkun
Reviewer: radix
Fixes [#10726](https://github.com/twisted/twisted/issues/10726)

This replaces one confusing, load-sensitive callLater test with several documented,
purposeful tests which should be immune to load-induced failures.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
